src/analytics/ledger/ledger_outputs.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(missing_docs)]
+
+use std::ops::{AddAssign, SubAssign};
+
+use derive_more::{AddAssign, SubAssign};
+
+use super::TransactionAnalytics;
+use crate::{
+    db::collections::analytics::LedgerOutputAnalyticsResult,
+    types::{
+        ledger::{LedgerOutput, LedgerSpent, MilestoneIndexTimestamp},
+        stardust::block::Output,
+    },
+};
+
+#[derive(Copy, Clone, Debug, Default, AddAssign, SubAssign)]
+pub struct CountValue {
+    pub count: usize,
+    pub value: u64,
+}
+
+impl AddAssign<&LedgerOutput> for CountValue {
+    fn add_assign(&mut self, rhs: &LedgerOutput) {
+        self.count += 1;
+        self.value += rhs.output.amount().0;
+    }
+}
+
+impl SubAssign<&LedgerSpent> for CountValue {
+    fn sub_assign(&mut self, rhs: &LedgerSpent) {
+        self.count -= 1;
+        self.value -= rhs.output.output.amount().0;
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct LedgerOutputAnalytics {
+    pub alias: CountValue,
+    pub basic: CountValue,
+    pub nft: CountValue,
+    pub foundry: CountValue,
+    pub treasury: CountValue,
+}
+
+impl LedgerOutputAnalytics {
+    /// Initialize the analytics by reading the current ledger state.
+    pub fn init<'a>(unspent_outputs: impl IntoIterator<Item = &'a LedgerOutput>) -> Self {
```

Comment:
I am taking this as the example ... I noticed that many analytics need an `init`ialization which begs the question whether we should make it part of the contract itself. I know it would require another associated type to account for the possibly different initialization data that is required, but I feel like that would be cleaner and force the implementor to think about whether he has prepared the analytics object correctly for the analytic to be correct. I am fine with the impl as it is right now, but just maybe something we want to revisit later.

Reply:
I am considering it., for sure.